src/implementations/orthnull.jl

```diff
@@ -5,287 +5,175 @@ copy_input(::typeof(right_orth), A) = copy_input(lq_compact, A) # do we ever nee
+
+initialize_output(::typeof(right_null!), A, alg::AbstractAlgorithm) =
+    initialize_output(right_null_kind(alg), A, alg)
+initialize_output(::typeof(right_null_lq!), A, alg::AbstractAlgorithm) =
+    initialize_output(lq_null!, A, alg)
+initialize_output(::typeof(right_null_svd!), A, alg::AbstractAlgorithm) = nothing
+
+function initialize_orth_svd(A::AbstractMatrix, F, alg)
+    S = Diagonal(initialize_output(svd_vals!, A, alg))
+    return F[1], S, F[2]
 end
+# fallback doesn't re-use F at all
+initialize_orth_svd(A, F, alg) = initialize_output(svd_compact!, A, alg)
 
 # Implementation of orth functions
 # --------------------------------
-function left_orth!(
-        A, VC;
-        trunc = nothing, kind = isnothing(trunc) ? :qr : :svd,
-        alg_qr = (; positive = true), alg_polar = (;), alg_svd = (;)
-    )
-    if !isnothing(trunc) && kind != :svd
-        throw(ArgumentError("truncation not supported for left_orth with kind=$kind"))
-    end
-    if kind == :qr
-        return left_orth_qr!(A, VC, alg_qr)
-    elseif kind == :polar
-        return left_orth_polar!(A, VC, alg_polar)
-    elseif kind == :svd
-        return left_orth_svd!(A, VC, alg_svd, trunc)
-    else
-        throw(ArgumentError("`left_orth!` received unknown value `kind = $kind`"))
-    end
-end
-function left_orth_qr!(A, VC, alg)
-    alg′ = select_algorithm(qr_compact!, A, alg)
-    check_input(left_orth!, A, VC, alg′)
-    return qr_compact!(A, VC, alg′)
-end
-function left_orth_polar!(A, VC, alg)
-    alg′ = select_algorithm(left_polar!, A, alg)
-    check_input(left_orth!, A, VC, alg′)
-    return left_polar!(A, VC, alg′)
-end
-function left_orth_svd!(A, VC, alg, trunc::Nothing = nothing)
-    alg′ = select_algorithm(svd_compact!, A, alg)
-    check_input(left_orth!, A, VC, alg′)
-    U, S, Vᴴ = svd_compact!(A, alg′)
-    V, C = VC
-    return copy!(V, U), mul!(C, S, Vᴴ)
-end
-function left_orth_svd!(A::AbstractMatrix, VC, alg, trunc::Nothing = nothing)
-    alg′ = select_algorithm(svd_compact!, A, alg)
-    check_input(left_orth!, A, VC, alg′)
-    V, C = VC
-    S = Diagonal(initialize_output(svd_vals!, A, alg′))
-    U, S, Vᴴ = svd_compact!(A, (V, S, C), alg′)
-    return U, lmul!(S, Vᴴ)
-end
-function left_orth_svd!(A, VC, alg, trunc)
-    alg′ = select_algorithm(svd_compact!, A, alg)
-    check_input(left_orth!, A, VC, alg′)
-    alg_trunc = select_algorithm(svd_trunc!, A, alg′; trunc)
-    U, S, Vᴴ = svd_trunc!(A, alg_trunc)
-    V, C = VC
-    return copy!(V, U), mul!(C, S, Vᴴ)
-end
-function left_orth_svd!(A::AbstractMatrix, VC, alg, trunc)
-    alg′ = select_algorithm(svd_compact!, A, alg)
-    check_input(left_orth!, A, VC, alg′)
-    alg_trunc = select_algorithm(svd_trunc!, A, alg′; trunc)
-    V, C = VC
-    S = Diagonal(initialize_output(svd_vals!, A, alg_trunc.alg))
-    U, S, Vᴴ = svd_trunc!(A, (V, S, C), alg_trunc)
-    return U, lmul!(S, Vᴴ)
-end
+left_orth!(A, VC, alg::AbstractAlgorithm) = left_orth_kind(alg)(A, VC, alg)
```

Comment:
What about something like this:
```julia
left_orth!(A, VC, alg::AbstractAlgorithm) = left_orth!(A, VC, left_orth_alg(alg))
```
and then there would be definitions like:
```julia
left_orth_alg(::Union{LAPACK_HouseholderQR, LAPACK_HouseholderQL}) = LeftOrthViaQR(alg)
```

Reply:
Or maybe that logic could be handled by definitions like:
```julia
select_algorithm(::typeof(left_orth!), A, alg::LAPACK_HouseholderQR) = LeftOrthViaQR(alg)
```
(I see now you said you didn't finish the implementation, partially I'm trying to anticipate what the full design might look like.)